### DISCUSSION TITLE
Parcel2: Prepending a remote URL to every asset

### DISCUSSION THREAD

**Discussion by MaffooBristol** [parcel-bundler/parcel#5732] [Q&A]
Following on from the conversation a while back on https://github.com/parcel-bundler/parcel/issues/4747

This issue was actually opened by a former colleague, and the temporary fix was to fork Parcel v1 and edit the source directly in order to do the loading as desired. This was done by modifying v1's `bundle-url/getBundleURL()` and `bundle-loader/loadBundle(bundle)` functions with some very specific and non-portable code.

Due to the fork being very out of date, and modules being taken down from npm, we're upgrading to Parcel 2. But, of course, the issue still remains - so I want to achieve this fix again, but done "properly".

Not 100% sure what the "proper" method is, but I assume it involves custom [runtimes](https://v2.parceljs.org/plugin-system/runtime/) or custom [transformers](https://v2.parceljs.org/plugin-system/transformer/). These seem to be not hugely documented for v2 at the moment, so I'd be grateful for any guidance. The **tl;dr** is as follows:

We are building our bundles. They work fine when we have control over how the HTML, JS and static files are loaded (ie, the "normal" way where the root JS file is loaded in and that relatively loads in all the assets within the same folder).

However we have integrated with a massive corporate CMS that we have 0 control over, and one of the ways it works is to treat all apps (such as ours) as a "widget" and is loaded in dynamically via their own Javascript. Therefore, our app will bootstrap fine, but the instant it tries to load in any assets or lazy-load via `require()`, it breaks. The only way we know to fix this is to give a url (say, `https://foo.com/`) and prepend that to EVERY reference to an asset.

The `--public-url` flag seems initially promising, but that only does it on the initial load. Eg,

```
<script src="./index.js" />
```

... becomes:

```
<script src="https://foo.com/index.js" />
```

But really we want every single JS asset file to do something like this:
```
// ...
"PEmA":[function(require,module,exports) {
module.exports="http://foo.com/icon-chevron-down-small-green.79eb1c01.svg";
},{}],"CQdS":[function(require,module,exports) {
module.exports="http://foo.com/icon-chevron-up-small-green.2c0f8a88.svg";
},{}],"ib1x":[function(require,module,exports) {
// ....
```

So the burning question is:

1. Can this be done natively as it stands in Parcel 2?
2. Is there a way this can be easily achieved using some form of plugin; if so, could you point me in the right direction as to how?
3. Is this something that should be added to Parcel?

Thanks in advance. Really appreciate it!
Matt

**Comment by MaffooBristol** [comment]
To my surprise and delight, it seems Parcel 2 actually does do this. It just applies it at runtime rather than putting it into the build output.

I was trying to write a runtime plugin and then tested and it used the public url as expected.

~So the answer is 1. yes, it can be done natively.~

**Reply by MaffooBristol** [reply]
I was wrong. This is still an issue. The `--public-url` causes the entry file to correctly prepend the initial JS bundle load. But subsequent assets ignore this value and try and load from the site root, which fails.

**Reply by MaffooBristol** [reply]
Maybe I can simplify the issue as steps:

`parcel build src/index.html`
Go to: http://example.com/long/path/to/subfolder/index.html
Tries to load http://example.com/index.hash.js
Fails

So we try:
`parcel build src/index.html --public-url http://example.com/long/path/to/subfolder/`
Go to: http://example.com/long/path/to/subfolder/index.html
Successfully loads http://example.com/long/path/to/subfolder/index.hash.js
However the index.hash.js file does a load of `import()`-based lazy loading and wants Login.hash.js
Rather than loading it at http://example.com/long/path/to/subfolder/Login.hash.js, it loads it at http://example.com/Login.hash.js
Mass of 404s ensues and breaks

The crux of the issue appears to be that it doesn't respect the public-url when loading anything other than the initial bundle defined in the entry point HTML.

@mischnic You've been really helpful recently with other parcel queries and I'm very thankful! But this is the "big one" that is blocking our releases which I incorrectly had thought had successfully been sorted by parcel2 out of the box, but was due to a pipeline issue that was running the old version. If you could advise and point me in the right direction, I'd be eternally grateful 🙏

**Reply by mischnic** [reply]
At the moment, this is how it works:
`import("./async.js")` becomes `loadViaScriptTag(getBundleURL() + relativePath("mainBundleId","asyncBundleId"))`

https://github.com/parcel-bundler/parcel/blob/7717e2547f510c89ea57bbfec62147bfb912ba49/packages/runtimes/js/src/bundle-url.js#L12-L30

So it should always use the current path of that bundle at runtime (regardless of what `---public-url` is)

**Reply by MaffooBristol** [reply]
Thanks @mischnic - I removed lazy loading from our build and now it doesn't brick, but the images don't load. So it's still an issue.

As previously mentioned, our bundle is loaded in using a kind of... widget-like functionality. I wonder whether this is causing "currentScript" to not be correct, or be blank, or other?

We had a hacked version of Parcel v1 which in fact edited that exact bit of code to put our URL in, but it's far from ideal. This Parcel v2 upgrade is due to our previous version no longer having access to certain npm packages which have since been removed (IIRC, `@parcel/babel-register`.

Since the current paradigm is to try and guess the correct path based on information provided, I think it would be good to be able to tell parcel to explicitly resolve to a URL (eg, the public-url)

Thanks!

**Reply by mischnic** [reply]
> currentScript

that isn't used, but instead the stacktrace is parsed. The last time I investigated currentScript, I think Safari was returning some strange value.

I've tried this:

```
{
	"dependencies": {},
	"targets": {
		"default": {
			"publicUrl": "/www",
			"distDir": "dist/www"
		}
	}
}
```
```js
import x from "url:./img.png";

console.log(x);
import("./async.js");
```
and then with `serve dist` and opening `http://localhost:5000/www/`, `http://localhost:5000/www/img.ecf8427e.png` correctly printed.

**Reply by MaffooBristol** [reply]
> that isn't used

Yep I missed the check for a test env

> I've tried this [...] correctly printed

Yeah, but I think the issue we're having is that the way some external stuff is loading/XHRing our code in, the current method of throwing + catching an error and parsing the stack trace isn't working.

I know that the resolution code in Parcel v1 and Parcel v2 both work for most normal situations, and that this is a bit of an edge case. But if that code were to do something like (pseudocode):

```
function getBundleURL() { 
  const publicUrl = getPublicUrl();
  if (publicUrl) {
    if (/^http/.test(publicUrl)) {
      return publicUrl;
    }
    return getBaseUrl() + publicUrl;
  }
  // ....
}
```

... would that cause any regressions? Basically, use the public URL if it's provided, otherwise fall back to the stack trace approach?

**Reply by mischnic** [reply]
I don't think your approach would work (without further changes).

Any chance you can provide a code sample of your situation? (where the async import loads the wrong url).

**Reply by MaffooBristol** [reply]
It's not just the lazy loading, it's all assets that need to be loaded outside of the main entry bundle. So, even with removing all lazy loading and async imports, it doesn't load any images/media from the correct subpath either.

There isn't really anything special code-wise to show either. I think it's just simply that it won't handle relative paths properly, and the only way to get it to work is to force it to always resolve absolutely.

All image imports are done in the normal way with `import img from './img.png` syntax. But I think that `getBundleURL()` can't find the relative path correctly.

We have some other builds from the same codebase which are self-hosted and served the normal way, with no issues. It's just this integration into a huge company's CMS that causes this issue.

Can `getBundleURL()` be overridden in a Parcel 2 custom plugin, if it couldn't be fixed in the main repo without major changes? If so, which plugin type/layer would that require?

**Reply by mischnic** [reply]
> isn't really anything special code-wise to show either

I asked because I still can't quite visualize your situation (= how the script is loaded at `/` but the images need to be loaded from `/subpath/...`).

> Can getBundleURL() be overridden in a Parcel 2 custom plugin

Yes, you would need to fork the `@parcel/runtime-js` plugin and override it in the parcelrc.

**Reply by MaffooBristol** [reply]
> how the script is loaded at `/`

- We are given a subpath on this big site. example.com/subpath/ourapp/
- We provide the CMS people with our bundle, which goes through a pipeline and goes onto an S3 bucket
- Their CMS injects our bundle's HTML into the page, and (I believe), a script on their site at example.com/loader.js, parses this injected code to find any `<script>` tags linking to JS files and then AJAXes it in with some jQuery.
- Therefore, although the page the user visits is example.com/subpath/ourapp/ - the script loading it lives at `/`, and therefore parcel sees it, does that throw+catch+trace thing, but as far as it knows, the script lives at `/`

This issue affects the main bundle as well as any other assets it tries to load. If we set a `public-url` then it fixes the initial load but then gets lost once it starts going into the territory of loading stuff relatively to the "current script", which Parcel expects to be the subpathed `/subpath/ourapp/index.hash.js`, but is in reality our client's `loader.js`

Does that make sense?

> Yes, you would need to fork the @parcel/runtime-js plugin and override it in the parcelrc.

Thanks!

**Reply by MaffooBristol** [reply]
Managed to sort it by basically taking our previous hack and putting it into a plugin instead, so thanks for the advice on that.

It's doing this not-so-great thing where we define a token in the `getBundleURL()` function and then do a big sed-replace in the dist folder as a build step to replace it with the public url. But of course, if this could come from the `--public-url` then it would be much less hacky.

In my pseudocode I call a fake function called `getPublicUrl()` - is there an actual proper way of doing this that wouldn't rely on trying to check the cli args? (I'm aware there are multiple ways of defining it)

Thanks for all your help today; been a lifesaver 🏅
